### PR TITLE
DAT-344 from https://github.com/GreaterLondonAuthority/Data_…

### PR DESCRIPTION
**Pull request by callum-oakley**
…for_London/pull/44

Created from https://github.com/GreaterLondonAuthority/Data_for_London/pull/44 by running

```
git format-patch --relative=src/ckanext-gla origin/develop --stdout > dat-344.mbox
```

in the main repository and then applying the patch to this repository with

```
git am ../../../Data_for_London-alt/dat-344.mbox
```

### CODE REVIEW

File: ckanext/gla/templates/snippets/package_list.html [hunk 1]
```diff
@@ -14,21 +14,29 @@
...
+            {% set is_search = h.is_search_results_page(request) %}
             {% if h.show_favourite_datasets(g.userobj, request) %}
                 {% set favourites = h.get_followed_datasets(g.userobj, request) %}
                 {% if favourites %}
                     {% snippet 'snippets/favourites.html', favourites=favourites %}
                 {% endif %}
             {% endif %}
             {% set nonfavourites = h.remove_favourites(g.userobj, request, packages) %}
-                <ul class="{{ list_class or 'dataset-list list-unstyled' }}">
+                <ul class="{{ list_class or 'dataset-list list-unstyled' }}" data-module="gla_search_result_logging">
```
Comment (kjoshi) [inline ckanext/gla/templates/snippets/package_list.html]: I'm not sure the `data-module` is doing anything here.`search_result_logging.js` doesn't start with: `ckan.module('gla_search_result_logging',...)` so I don't think a module with that name gets registered with CKAN.

I don't think the module is really necessary anyway if you're just calling `logResultClick` directly in `package_item`
```suggestion
                <ul class="{{ list_class or 'dataset-list list-unstyled' }}">
```

File: ckanext/gla/templates/admin/index.html [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+        {% set docs_url = "http://docs.ckan.org/en/{0}/sysadmin-guide.html".format(g.ckan_doc_version) %}
+        {% trans %}
+        <p>As a sysadmin user you have full control over this CKAN instance. Proceed with care!</p>
+        <p>For guidance on using sysadmin features, see the CKAN  <a href="{{ docs_url }}" target="_blank" rel="noreferrer">sysadmin guide</a></p>
+        {% endtrans %}
+    </div>
+    <div class="module-content">
+        <p>System administrators can download a log of all user searches, and the results that they selected
+            as a csv file <a id="downloadlink"  data-module="gla_search_result_log_download">here</a>.</p>
```
Comment (kjoshi) [inline ckanext/gla/templates/admin/index.html]: Similar comment to `package_item.html` - I don't think the `data-module` is doing anything here.
```suggestion
            as a csv file <a id="downloadlink">here</a>.</p>
```

File: ckanext/gla/search.py [hunk 3]
```diff
@@ -34,3 +41,36 @@ def debug(context, data_dict={}):
         return conn.search(**params).__dict__
     except Exception as e:
         raise common.SearchError(e.args)
+
+logfile = "/srv/app/search_logs.csv"
+
+def _result_index(page, index_in_page):
+    page_idx = 0 if _empty_or_none(page) else int(page) - 1
+    return page_idx * 20 + int(index_in_page) + 1
```
Comment (kjoshi) [inline ckanext/gla/search.py]: The number of datasets per page can be set in the [config](https://github.com/ckan/ckan/blob/master/ckan/config/config_declaration.yaml#L1078), so it might be best to link this to that config value in case it get's changed at some point?

Also, according to the docs `loop.index` (in `package_list.html`) starts at 1 already:
image
so I'm pretty sure we don't need the extra `+1` here

```suggestion
    from ckan.common import config
    return page_idx * config.get('ckan.datasets_per_page') + int(index_in_page)
```

**Review by kjoshi — CHANGES_REQUESTED** [review]
I think I’ve found a bit of a nicer way of handling the file download.
I’ve created a new branch with some [example code](https://github.com/GreaterLondonAuthority/dfl-ckanext/commit/cfcd571bdfab6a1fe49694875a21178f56a759c2).

TLDR: We need a CKAN `view` to serve the file (by using `flask.send_file`), rather than an `action`.
My example code is very rough but hopefully you can see the idea. 
It would mean that the `get_logs` action could be replaced by the new view function and the `<script>` in `admin/index.html` could just be replaced with a link to the new route.

If you think this seems like a good idea maybe you could incorporate it into this PR? 

Another idea I had was to make the log file path configurable using [IConfigDeclaration](https://docs.ckan.org/en/2.10/extensions/plugin-interfaces.html#ckan.plugins.interfaces.IConfigDeclaration) ([example here](https://docs.ckan.org/en/2.10/maintaining/configuration.html#declaring-config-options)) so that it's easier to get access to that variable in different files, and maybe easier to change in future. (E.g. we could create a new `ckanext.gla.search_logs_path` config variable)

But that could be overkill if this isn't a permanent solution...

**Review by kjoshi — APPROVED** [review]
Looks good to me! 👍
